src/dataflow/src/render/debezium.rs

```diff
@@ -118,6 +129,259 @@ where
+                activator.activate().unwrap()
+            }
+        }
+    });
+
+    // XXX(chae): use an activator to ensure progress?
+    // Current state: only `max(Progress.upper) - 1` is considered closed by the coordinator.  So things written at max cannot be queried via sql.
+    tx_ok.inner.sink(Pipeline, "envelope-debezium-txdata", {
+        let channel = Rc::clone(&channel);
+        let activator = Rc::clone(&activator);
+        let tx_map = Rc::clone(&tx_map);
+        //let count = Rc::clone(&count);
+        //let mut dedup_state = HashMap::new();
+        let mut data = vec![];
+        use timely::progress::Timestamp as _;
+        let last = Rc::new(RefCell::new(<G as ScopeParent>::Timestamp::minimum()));
+        move |input| {
+            while let Some((cap, refmut_data)) = input.next() {
+                refmut_data.swap(&mut data);
+                for (row, time, diff) in data.drain(..) {
+                    assert_eq!(diff, 1);
+                    let mut i = row.iter();
+                    let (status, tx_id, event_count): (
+                        _,
+                        <G as ScopeParent>::Timestamp,
+                        Option<i64>,
+                    ) = (
+                        i.next().unwrap().unwrap_str(),
+                        i.next().unwrap().unwrap_str().parse().unwrap(),
+                        i.next().unwrap().try_into().unwrap(),
+                    );
+                    eprintln!(
+                        "TX META STREAM: {:?} {:?} {:?}; TIME: {:?}; CAP TIME {:?}",
+                        status,
+                        tx_id,
+                        event_count,
+                        time,
+                        cap.time(),
+                    );
+                    tx_map.borrow_mut().insert(tx_id, time);
+                    if status == "END" {
+                        let data_time = tx_id;
+                        let tx_upper_time = tx_id + 1;
+                        // XXX(chae): Transform into time domain provided by the tx data rather than using the tx_id as the time
+                        channel.borrow_mut().push_back(Message::Progress(Progress {
+                            lower: vec![last.borrow().clone()],
+                            upper: vec![(tx_upper_time.clone())],
+                            counts: vec![(data_time.clone(), event_count.unwrap() as usize)],
+                        }));
+                        *last.borrow_mut() = tx_upper_time;
+                    }
+                }
+            }
+            if let Some(activator) = activator.borrow_mut().as_mut() {
+                activator.activate().unwrap()
+            }
+        }
+    });
+    struct VdIterator<T>(Rc<RefCell<VecDeque<T>>>);
+    impl<T: std::fmt::Debug> Iterator for VdIterator<T> {
+        type Item = T;
+        fn next(&mut self) -> Option<T> {
+            eprintln!("PROVIDING {:?}", self.0.borrow_mut().get(0));
```

Comment:
for `debug` messages, you might prefer the `dbg!` macros, which lets you do stuff like
```
return dbg!(self.0.borrow_mut().pop_front())
```
and it passes the value through, and prints the expression that caused it and line number

Reply:
good tip!